staging/vhost-device-gpu/src/virt_gpu.rs

```diff
@@ -330,16 +341,36 @@ impl VirtioGpu {
     }
 
     /// If the resource is the scanout resource, flush it to the display.
-    pub fn flush_resource(&mut self, resource_id: u32) -> VirtioGpuResult {
+    pub fn flush_resource(&mut self, resource_id: u32, gpu_backend: &mut GpuBackend, mem: &GuestMemoryMmap) -> VirtioGpuResult {
         if resource_id == 0 {
             return Ok(OkNoData);
         }
 
-        #[cfg(windows)]
-        match self.rutabaga.resource_flush(resource_id) {
-            Ok(_) => return Ok(OkNoData),
-            Err(RutabagaError::Unsupported) => {}
-            Err(e) => return Err(ErrRutabaga(e)),
+        let resource = self.resources.get(&resource_id).unwrap();
+        if resource.scanout_width != 0 && resource.scanout_height != 0 {
+            debug!("flush resource with associated scanout: {}x{}",resource.scanout_width, resource.scanout_height);
+            // lets assume hardcoded XRGB
+            const BYTES_PER_PIXEL: usize = 4;
+            let size = BYTES_PER_PIXEL * resource.scanout_width as usize * resource.scanout_height as usize;
+            let mut data = vec![0u8; size];
+            let mut pos = 0;
+            for (addr, len) in resource.backing_iovecs.as_ref().unwrap().iter().copied() {
+                debug!("iovec: {}..{}", pos, pos+ len);
+                let bytes_read = mem.read(&mut data[pos..pos+len], addr).unwrap();
+                assert_eq!(bytes_read, len);
+                pos += bytes_read;
```

Comment:
```suggestion
                pos += bytes_read;
```
                // Ensure that pos + len does not exceed the length of data
                if pos + len > data.len() {
                    // Adjust len to prevent out-of-bounds access
                    let remaining_len = data.len() - pos;
                    let bytes_read = mem.read(&mut data[pos..], addr).unwrap();
                    assert_eq!(bytes_read, remaining_len);
                    pos += bytes_read;
                } else {
                    let bytes_read = mem.read(&mut data[pos..pos+len], addr).unwrap();
                    assert_eq!(bytes_read, len);
                    pos += bytes_read;
                }